### ISSUE TITLE
E/AndroidRuntime: FATAL EXCEPTION: Thread-2

### ISSUE THREAD

**Issue by semw** [schwabe/ics-openvpn#1239]
To make issues more manageable, I would appreciate it if you fill out the following details as applicable:

# General information
1. Android Version=10
2. Android Vendor/Custom ROM=AOSP built from source code
3. Device= Google Pixel 2XL
4. Version of the app (version number/play store version/self-built) =Ics-openvpn and OpenVPN for android from this [link](https://openvpn-for-android.en.uptodown.com/android/download)
 
# Description of the issue
I have added pre-built apk to AOSP source code and built it. I have added this app libraries into AOSP source code as well. After compiling aosp this app crashes on connect or it keeps busy on building configuration. I have added other apps with libraries , they work fine but ics-openvpn or OpenVPN for android does not work.

I have checked [this](https://github.com/schwabe/ics-openvpn/issues/662) thread of issue as well but I have already added libraries to my AOSP

Below are my logs of app
Any help will be highly appreciated

`  --------- beginning of crash
2020-09-24 16:55:31.201 6433-6497/? E/AndroidRuntime: FATAL EXCEPTION: Thread-2
    Process: de.blinkt.openvpn:openvpn, PID: 6433
    java.lang.UnsatisfiedLinkError: No implementation found for java.lang.String de.blinkt.openvpn.core.NativeUtils.getJNIAPI() (tried Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI and Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI__)
        at de.blinkt.openvpn.core.NativeUtils.getJNIAPI(Native Method)
        at de.blinkt.openvpn.core.NativeUtils.getNativeAPI(NativeUtils.java:25)
        at de.blinkt.openvpn.VpnProfile.getPlatformVersionEnvString(VpnProfile.java:684)
        at de.blinkt.openvpn.VpnProfile.getConfigFile(VpnProfile.java:354)
        at de.blinkt.openvpn.VpnProfile.writeConfigFile(VpnProfile.java:774)
        at de.blinkt.openvpn.core.OpenVPNService.startOpenVPN(OpenVPNService.java:586)
        at de.blinkt.openvpn.core.OpenVPNService.access$200(OpenVPNService.java:66)
        at de.blinkt.openvpn.core.OpenVPNService$3.run(OpenVPNService.java:565)
        at java.lang.Thread.run(Thread.java:922)
2020-09-24 16:55:32.456 6504-6504/? E/openvpn:openvp: No implementation found for java.lang.String de.blinkt.openvpn.core.NativeUtils.getJNIAPI() (tried Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI and Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI__)
2020-09-24 16:55:32.476 6504-6528/? E/openvpn:openvp: No implementation found for java.lang.String de.blinkt.openvpn.core.NativeUtils.getJNIAPI() (tried Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI and Java_de_blinkt_openvpn_core_NativeUtils_getJNIAPI__)
2020-09-24 16:55:32.476 6504-6528/? E/AndroidRuntime: FATAL EXCEPTION: Thread-2`

**Comment by schwabe** [comment]
You seem to be missing the native libraries or the are at a place where they cannot be loaded. You should already gotten an error/warning on the `System.loadLibrary("opvpnutil");`. The other APKs you tried probably did not have .so libraries. 

I am not sure what the bigger picture is you try to achieve here. Preinstall the app? Somehow integrate it deep into the ROM?

**Comment by semw** [comment]
I have placed libraries in OS before building it. I also verified after flashing AOSP (OS) in device, my libraries are present in system/lib folder. I used same process for other apps and their libraries are also present in system/lib folder. Other apps which i have added do have dependent libraries.
And yes I am making it as pre-built app (System app) like device manufacturer put their apps in OS which cannot be deleted and also remain in OS even after factory reset.
Further I am not getting error/warning on the `System.loadLibrary("opvpnutil");`
